compiler/src/dotty/tools/backend/sjs/JSExportsGen.scala

```diff
@@ -0,0 +1,914 @@
+        js.Undefined()
+      }
+    } else {
+      genApplyMethod(targetTree, defaultGetter, defaultGetterArgs)
+    }
+  }
+
+  private def targetSymForDefaultGetter(sym: Symbol): Symbol = {
+    if (sym.isClassConstructor) {
+      /*/* Get the companion module class.
+       * For inner classes the sym.owner.companionModule can be broken,
+       * therefore companionModule is fetched at uncurryPhase.
+       */
+      val companionModule = enteringPhase(currentRun.namerPhase) {
+        sym.owner.companionModule
+      }
+      companionModule.moduleClass*/
+      sym.owner.companionModule.moduleClass
+    } else {
+      sym.owner
+    }
+  }
+
+  private def defaultGetterDenot(targetSym: Symbol, sym: Symbol, paramIndex: Int): Denotation =
+    targetSym.info.member(DefaultGetterName(sym.name.asTermName, paramIndex))
+
+  private def defaultGetterDenot(sym: Symbol, paramIndex: Int): Denotation =
+    defaultGetterDenot(targetSymForDefaultGetter(sym), sym, paramIndex)
+
+  /** Generate the final forwarding call to the exported method. */
+  private def genResult(exported: Exported, args: List[js.Tree], static: Boolean)(
+      implicit pos: SourcePosition): js.Tree = {
+
+    val sym = exported.sym
+
+    def receiver = {
+      if (static)
+        genLoadModule(sym.owner)
+      else if (sym.owner == defn.ObjectClass)
+        js.This()(jstpe.ClassType(ir.Names.ObjectClass))
+      else
+        js.This()(encodeClassType(sym.owner))
+    }
+
+    def boxIfNeeded(call: js.Tree): js.Tree = {
+      box(call, atPhase(elimErasedValueTypePhase)(sym.info.resultType))
+    }
+
+    if (currentClassSym.isNonNativeJSClass) {
+      assert(sym.owner == currentClassSym.get, sym.fullName)
+      boxIfNeeded(genApplyJSClassMethod(receiver, sym, args))
+    } else {
+      if (sym.isClassConstructor)
+        js.New(encodeClassName(currentClassSym), encodeMethodSym(sym), args)
+      else if (sym.isPrivate)
+        boxIfNeeded(genApplyMethodStatically(receiver, sym, args))
+      else
+        boxIfNeeded(genApplyMethod(receiver, sym, args))
+    }
+  }
+
+  private def genThrowTypeError(msg: String = "No matching overload")(implicit pos: Position): js.Tree =
+    js.Throw(js.JSNew(js.JSGlobalRef("TypeError"), js.StringLiteral(msg) :: Nil))
+
+  private final class ParamSpec(val sym: Symbol, val info: Type,
+      val isRepeated: Boolean, val hasDefault: Boolean) {
+    override def toString(): String =
+      s"ParamSpec(${sym.name}, ${info.show}, isRepeated = $isRepeated, hasDefault = $hasDefault)"
+  }
+
+  private object ParamSpec {
+    def apply(methodSym: Symbol, sym: Symbol, infoAtElimRepeated: Type, infoAtElimEVT: Type,
+        methodHasDefaultParams: Boolean, paramIndex: Int): ParamSpec = {
+      val isRepeated = infoAtElimRepeated.isRepeatedParam
+      val info = if (isRepeated) infoAtElimRepeated.repeatedToSingle else infoAtElimEVT
+      val hasDefault = methodHasDefaultParams && defaultGetterDenot(methodSym, paramIndex).exists
```

Comment:
In dotty, params themselves are not marked as default parms. The *method* is marked as having one or more defaults. This is what `methodHasDefaultParams` tests. Then, once the the method has default params, we need to test whether a particular parameter has a default value by testing whether a default getter exists for it, depending on its owning method name and its index.

Both tests are necessary. If we only test whether a default getter exists, we would mistakenly think that a parameter has a default value, when in fact it is a parameter in the same position *of an overload* of the method that has a default value, like:
```scala
def foo(a: Int, x: Int): Int = ???
def foo(b: Int, y: String = "hello"): String = ???
```
we would think that `x` has a default value, because `defaultGetterDenot` would return the getter for `"hello"`. And if we only test `methodHasDefaultParams`, we would think that `b` has a default value.

So both sides of the `&&` are required and make sense.